core(flex): Flex layout primitive

JonasBa opened the pull request:
_The PR modifies the Flex component to build on top of the newly added Container component and allow for a higher degree of flexibility via responsive props._

Our motivation is to reduce the amount of styled components we have to write and streamline layout operations in the app. This PR follows the recently added [Text](https://sentry.sentry.io/stories/core/text) primitive, and is the first primitive that we are introducing under our layout primitives with other layouts like Grid or Stack to follow next. 

I am opening this as a draft with the intention of collecting feedback and settling on the most ergonomic solution. The changes and approach we settle on here, will serve as the baseline for other components, so it is important to me that we find a good layout API.

**API Notes:**
- `Flex` builds on top of a Container and inherits is properties. It is the job of the container to support non flex related layout properties.

- I have intentionally avoided calling the `Container` a `Box` due to the baggage it comes with in the context of Sentry - we've had this once already and removed it, so I don't want anyone to think we are bringing it back.

- Container is exported and could be consumed. I am unsure if this is a good idea and I may actually remove it. 
The drawback is that Container currently supports margins to be specified, which could become a source of future problems. We want to discourage margins in components, and the best way to do that is to simply not do that at all. A popular approach among layout primitives in other design systems seems to be to simply not allow setting margins, and force users to either use gap or padding.

- We use responsive props to manage media queries. There are many API versions that implement this differently, all of which have their own drawbacks that I will try to explain.

	First, a responsive prop is a progressive enhancement of a static value, meaning it is of type of `T |  Record<'xs', 'sm' ...., T>`, which enables us to support defaults via a simple string prop while also being to support breakpoint specific values.

	Static gap:
	```tsx
	// sm on all sizes
	<Flex gap="sm">...</Flex>
	```
	Responsive gap:
	```tsx
	// 0 -> sm = xs gap
	// sm -> md = xs gap
	// md+ = md gap
	<Flex gap={{sm: 'xs', md: 'md'}}>...</Flex>
	```
	
	The other popular approach to implement this could be by using a reserved default.
	```tsx
	<Flex gap={{sm: 'xs', default: 'md'}}>...</Flex>
	```
	While this works, it feels redundant or verbose, and I would like to stretch ourselves to think of a better default. When selecting values for responsive design, the mental model is usually that there is some "default", or a "max", and some "min" that you want to use.
	
	An example of this is something like the flex direction for example, where a common case is to use a vertical layout on smaller screens.
	```tsx
	<Flex direction={{sm: 'column', md: 'row'}}>...</Flex.
	// ^ Should this use row for xs layout, or should we require an explicit value?
	
	// Both the option of requiring an explicit value and a default feel unnecessarily verbose to me.
	<Flex direction={{xs: 'column', sm: 'column', md: 'row'}}>...</Flex>
	<Flex direction={{default: 'row', xs: 'column', sm: 'column'}}>...</Flex>
	```
	The alternative idea that I am proposing is to treat the breakpoints like a min and max range, with individual value inbetween that allow you to customize the appearance if you need to, but do not require you to specify the values unless you want to opt out of this mechanism betting on the fact that majority probably don't want to do that.
	
	```tsx
	<Flex direction={{sm: 'column', md: 'row'}}>...</Flex>
	^ Treats column as the layout from 0-> md breakpoint, and row for md+ layouts
	```
	By doing this, we essentially defer the question of a `default` value selection to the user via a min and max while still providing an escape. The hypothesis here is that this would be an exception, and not something that developers will often need to use.
	```tsx
	<Flex direction={{sm: 'row' sm: 'column', md: 'row'}}>...</Flex>
	^ This should not be something that is used often, so we optimize for the alternative
	```

- Container queries are currently not supported yet. We had a chat with @TkDodo about making container queries the default, but I am not convinced that that is a good idea. One idea could be to expose them via a breakpoint prefix or a [container](https://developer.mozilla.org/en-US/docs/Web/CSS/container) boolean/string prop, which would allow us to target different containers via container name query and by extending the Responsive type prop

	```tsx
	<Flex container='parent'>
       ...
    	  <Flex direction={{sm: 'row' sm: 'column', md: 'row', container: 'parent'}}>...</Flex>
       ...
	</Flex>
	```

JonasBa committed:
core(flex): add container

JonasBa committed:
flex: add responsive props

JonasBa committed:
core(flex): better docs

JonasBa committed:
core(flex): better docs

JonasBa committed:
core(flex): improve docs

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+import {isChonkTheme} from 'sentry/utils/theme/withChonk';
+
+const BREAKPOINT_ORDER: readonly Breakpoint[] = ['xs', 'sm', 'md', 'lg', 'xl'];
+// We alias None -> 0 to make it slighly more terse and easier to read.
+type RadiusSize = keyof DO_NOT_USE_ChonkTheme['radius'] | '0';
+type SpacingSize = keyof Theme['space'] | '0';
```
I think this might help yeah, the 0 would be a bit more friendlier and less to write for something that could be a very common thing that people type, so for example something like
`<Flex padding="md none"/>` vs `<Flex padding="md 0"/>`

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+  | `${RadiusSize} ${RadiusSize}`
+  | `${RadiusSize}`;
+
+type Responsive<T> = T | Record<Breakpoint, T | undefined>;
```
Ah yes, will do. I've just been writing this in stories, so I haven't seen the error

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+  ${p => responsiveCSS('padding-left', p.pl, p.theme, getSpacing)};
+  ${p => responsiveCSS('padding-right', p.pr, p.theme, getSpacing)};
+
+  ${p => responsiveCSS('margin', p.m, p.theme, getSpacing)};
+  ${p => responsiveCSS('margin-top', p.mt, p.theme, getSpacing)};
+  ${p => responsiveCSS('margin-bottom', p.mb, p.theme, getSpacing)};
+  ${p => responsiveCSS('margin-left', p.ml, p.theme, getSpacing)};
+  ${p => responsiveCSS('margin-right', p.mr, p.theme, getSpacing)};
```
Oh yeah, this needs to respect the prop order 100%

JonasBa replied:
Given the codebase usage, we probably want to optimize for shorthand vs top/bottom, so if we do go for one, that one would be my preference.

JonasBa committed:
flex: improve types

JonasBa committed:
flex: use test.each in tests

TkDodo reviewed (commented):
regarding responsiveness: I still think the easiest would be to have breakpoints inherit from smaller values.

In the flex docs, we have:

> If a prop is not specified for a breakpoint, the value will **not** be inherited from the previous breakpoint.

which isn’t entirely true. From your example:

```
 <Flex direction={{sm: 'column', md: 'row'}}>...</Flex.
 // ^ Should this use row for xs layout, or should we require an explicit value?
```

Here, `lg` and `xl` breakpoints will inherit from `md`, right? So if we embrace the breakpoint order an have things in between also inherit, we could do:

```
 <Flex direction={{xs: 'column', md: 'row'}}>...</Flex.
```

This says:
- for xs, use column
- sm, since it’s between xs and md, it would also get column
- md and above get row

In other words, these are ranges:

from xs to md, we want column
from md upwards, we want row

I think this would be quite consistent.

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+type SpacingSize = keyof Theme['space'] | '0';
+type Breakpoint = keyof Theme['breakpoints'];
+
+// @TODO(jonasbadalic): audit for memory usage and linting performance issues.
+// These may not be trivial to infer as we are dealing with n^4 complexity
+type Spacing =
+  | `${SpacingSize} ${SpacingSize} ${SpacingSize} ${SpacingSize}`
+  | `${SpacingSize} ${SpacingSize} ${SpacingSize}`
+  | `${SpacingSize} ${SpacingSize}`
+  | `${SpacingSize}`;
```
Yeah, I was just thinking of the typeahead, but maybe we just define this a string?

TkDodo replied:
@natemoo-re that’s a good idea. with tuples, we only have a very small union type and we also wouldn’t need to split the string at the blank at runtime, but just need an array check. Also, no interpolation:

```
type Spacing =
  | [SpacingSize, SpacingSize, SpacingSize, SpacingSize]
  | [SpacingSize, SpacingSize, SpacingSize]
  | [SpacingSize, SpacingSize]
  | SpacingSize;
```

JonasBa replied:
The only problem with tuples is that they are a lot more cumbersome to write and edit. In this small example, the difference is 12 char vs 7, but what makes this more annoying is that all of these are special characters (2 square brackets, 2 extra dashes and a comma), which makes it cumbersome to write out
```tsx
"sm md"
["sm", "md"]
```

I think I would prefer to keep a string literal here, as I think the developer experience of it will be much nicer, but maybe we find a middle ground where we just type it as `${string}` | `${string} ${string}...` instead of the actual spacing sizes.

I would also like to understand what the performance overhead is here, before we decide to scrap the type, because it might be negligible and not even worth optimizing for

JonasBa committed:
flex: cleanup types and destructuring children

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+}
+/* eslint-enable typescript-sort-keys/interface */
+
+type ContainerElement =
+  | 'article'
+  | 'aside'
+  | 'div'
+  | 'figure'
+  | 'footer'
+  | 'header'
+  | 'li'
+  | 'main'
+  | 'nav'
+  | 'ol'
+  | 'section'
+  | 'span'
+  | 'summary'
+  | 'ul';
```
I think we would probably want a composition pattern here with Text as='p', but the API of flex doesn't support that yet

JonasBa replied:
I'm going to omit it for now, and work on finding a composition pattern here

natemoo-re commented on static/app/views/settings/organizationTeams/teamMembers.tsx:
```diff
@@ -374,7 +374,7 @@ function TeamMembers({team}: TeamMembersProps) {
             <TeamRoleColumnLabel />
           </div>
-          <Flex justify="end">
+          <Flex justify="flex-end">
```
I'm guessing it comes with the generalized `Container` type, but I agree that it feels duplicative.

TkDodo commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+]);
+
+export const Container = styled(
+  <T extends ContainerElement = 'div'>(props: ContainerProps<T>) => {
+    const Component = (props.as ?? 'div') as T;
+    return <Component {...(props as any)}>{props.children}</Component>;
+  },
```
yeah you would need to hoist the type assertion too:

```
as: Component = 'div' as T,
```

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+  ${p => responsiveCSS('min-height', p.minHeight, p.theme, v => v)};
+  ${p => responsiveCSS('max-height', p.maxHeight, p.theme, v => v)};
+
+  /**
+   * This cast is required because styled-components does not preserve the generic signature of the wrapped component.
+   * By default, the generic type parameter <T> is lost, so we use 'as unknown as' to restore the correct typing.
+   * https://github.com/styled-components/styled-components/issues/1803
+   */
+` as unknown as <T extends ContainerElement = 'div'>(
+  props: ContainerProps<T>
+) => React.ReactElement;
```
I tried adding it, and maybe I'm doing something wrong, but it seems like generics on styled components like these might be somewhat annoying here. The test that comes to mind would be to check that `ContainerProps<span>` results in `{ref: React.Ref<HTMLSpanElement>`, but we can test that via something like
```tsx
 const ref = createRef<HTMLOListElement>();
 render(<Component ref={ref} as="ol"/>);
```

JonasBa commented:
> regarding responsiveness: I still think the easiest would be to have breakpoints inherit from smaller values.
> 
> In the flex docs, we have:
> 
> > If a prop is not specified for a breakpoint, the value will **not** be inherited from the previous breakpoint.
> 
> which isn’t entirely true. From your example:
> 
> ```
>  <Flex direction={{sm: 'column', md: 'row'}}>...</Flex.
>  // ^ Should this use row for xs layout, or should we require an explicit value?
> ```
> 
> Here, `lg` and `xl` breakpoints will inherit from `md`, right? So if we embrace the breakpoint order an have things in between also inherit, we could do:
> 
> ```
>  <Flex direction={{xs: 'column', md: 'row'}}>...</Flex.
> ```
> 
> This says:
> 
> * for xs, use column
> * sm, since it’s between xs and md, it would also get column
> * md and above get row
> 
> In other words, these are ranges:
> 
> from xs to md, we want column from md upwards, we want row
> 
> I think this would be quite consistent.

I realized I did not explain that well, but omitted values inherit the values from the previous breakpoint (due to the use of min-width), so this already works like it.

JonasBa committed:
flex: revert flex change

JonasBa committed:
Merge branch 'master' into jb/core/flex-mdx

gggritso commented:
Very nice! Just a few comments from me, nothing major!

> `Flex` builds on top of a Container and inherits is properties. It is the job of the container to support non flex related layout properties.

This is interesting! I'm not opposed to this at all, but I think it's an interesting choice, since it might also be good to compose components instead:

```jsx
<Container p="sm">
  <Flex gap="sm"...
</Container>
```

> Container is exported and could be consumed.

I don't mind this, personally, especially because of the composition argument above (though I'd understand if you don't like that idea)

> The alternative idea that I am proposing is to treat the breakpoints like a min and max range

Fully in favour of this 👍🏻 This matches how I think about breakpoints. 

> Container queries are currently not supported yet

Fine by me! I wouldn't even be mad if there wasn't a primitive for this. Are there many uses of container queries in the codebase? Big Number widgets is one off the top of my head.

JonasBa committed:
flex: remove pb, pt, pl, pr

JonasBa committed:
flex: add order prop to flex

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,30 +1,343 @@
+  mb?: Responsive<Spacing>;
+  ml?: Responsive<Spacing>;
+  mr?: Responsive<Spacing>;
+  mt?: Responsive<Spacing>;
```
Lets just go with padding shorthand to start with and see how it goes in practice. We could add the other props later if we start to feel a need for it

JonasBa committed:
flex: remove 0 alias

TkDodo commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,38 +1,325 @@
+  'flex',
+  'gap',
+  'inline',
+  'align',
+  'justify',
```
I’m a bit confused about why we have `align` instead of `align-items` and `justify` instead of `justify-content`. Is it because it’s influenced by `direction` ? Like, do we get `align-items` or `align-content` dependent on `direction`, because I don’t think we do ...

I’m not really a fan of introducing new APIs unless necessary or clearly beneficial. The problem is that props aren’t discoverable because we allow all html-props, too, so there’s a lot of them. And anything that’s different from standard flex naming needs to be learned anew...

JonasBa replied:
Resolved in DM, but it seems that this is already handled by the layout engine when direction is column, nice find @TkDodo 🙏🏼

JonasBa commented on static/app/components/core/layout/flex.tsx:
```diff
@@ -1,38 +1,325 @@
+  ${p => rc('overflow-x', p.overflowX, p.theme)};
+  ${p => rc('overflow-y', p.overflowY, p.theme)};
+
+  ${p => rc('padding', p.padding, p.theme, getSpacing)};
```
I had solved the ordering, but I think it was solving something that didn't need to exist in the first place. If we remove it in favor of just using the padding prop, that should be fine. 

One of the problems with those props too is that they will inevitably end up not being colocated and easy for people to glance over. If you imagine something like this, the pb doesn't really stand out there in the end
```tsx
<Flex padding="sm md" justify="center" align="start" pb="xs"/>
// ^ pb and padding are in different places
<Flex padding="sm md xs md"/>
// ^ we force folks to modify the shorthand, keeping these in the same place
```

JonasBa committed:
core(flex): update flex api and remove gap space instances (#96324)

Align justify with flex and use `start` as opposed to `flex-start` (flex
prefix is redundant in the context of a flex component). Also removed a
couple instances of `gap={space}` that have been left over.

Note that this is a stacked PR and isn't getting merged into master as
it requires the core flex changes from
https://github.com/getsentry/sentry/pull/96206

JonasBa committed:
Merge branch 'master' into jb/core/flex-mdx

JonasBa committed:
flex: use space.0 and add grid-area to container

JonasBa committed:
Merge branch 'master' into jb/core/flex-mdx

JonasBa committed:
fix two old properties

JonasBa committed:
add usage of container
